File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/CallsHandler.java
package dev.alenajam.opendialer.feature.inCall.service;

import android.content.Context;
import android.telecom.Call;
import android.telecom.CallAudioState;

import androidx.annotation.MainThread;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

import dev.alenajam.opendialer.feature.inCall.R;
import dev.alenajam.opendialer.feature.inCall.ui.InCallActivity;

@Singleton
public class CallsHandler {
    private final MutableLiveData<Map<Call, OngoingCall>> calls = new MutableLiveData<>(new HashMap<>());
    private final MutableLiveData<CallDisplayState> displayState =
            new MutableLiveData<>(new CallDisplayState(null, null));
    private final MutableLiveData<CallAudioUiState> audioState = new MutableLiveData<>();
    private final MutableLiveData<Boolean> canAddCall = new MutableLiveData<>();
    private InCallServiceImpl callService;
    private InCallActivity inCallActivity;

    private Context context;
    private ProximitySensor proximitySensor;
    private long nextCallSequence;
    private final CallContactResolver contactResolver;

    @Inject
    public CallsHandler(CallContactResolver contactResolver) {
        this.contactResolver = contactResolver;
    }

    public void setInCallActivity(InCallActivity inCallActivity) {
        this.inCallActivity = inCallActivity;
    }

    public void clearInCallActivity(InCallActivity inCallActivity) {
        if (inCallActivity == this.inCallActivity) {
            this.inCallActivity = null;
        }
    }

    public boolean isActivityStarted() {
        return inCallActivity != null && !inCallActivity.isDestroyed() && !inCallActivity.isFinishing();
    }

    public boolean isActivityShowing() {
        if (!isActivityStarted()) return false;

        return inCallActivity.getVisibility();
    }

    @MainThread
    public void addCall(Call call, Context context) {
        if (call.getState() == Call.STATE_DISCONNECTED) {
            OngoingCallHelper.handleDisconnectCause(context, call);
            return;
        }

        OngoingCall ongoingCall = new OngoingCall(context, call, this, nextCallSequence++);

        // Copy the map so a new reference is posted. LiveData.postValue with the
        // same mutated-in-place HashMap instance is not detected as a change by
        // Compose's observeAsState (structural equality on the same reference
        // is always true), so conference UI would not recompose on add/remove.
        Map<Call, OngoingCall> map = new HashMap<>(calls.getValue());
        map.put(call, ongoingCall);

        // Telecom and Call callbacks run on the main thread. Publish immediately
        // so simultaneous conference updates cannot overwrite one another.
        calls.setValue(map);
        resolveContact(ongoingCall);
        updateCalls();
    }

    @MainThread
    public void removeCall(Call call) {
        if (calls.getValue() == null) return;

        Map<Call, OngoingCall> map = new HashMap<>(calls.getValue());

        OngoingCall ongoingCall = map.remove(call);
        if (ongoingCall == null) return;

        ongoingCall.tearDown();

        // Conference members can disconnect back-to-back. Using postValue here
        // lets each callback copy the same stale map, so the last posted removal
        // can restore calls removed by earlier callbacks.
        calls.setValue(map);
        updateCalls();
    }

    @MainThread
    public void updateCalls() {
        Map<Call, OngoingCall> map = calls.getValue();
        if ((map.isEmpty() || getPrimaryCallToDisplay() == null) && reconcileCallsFromTelecom()) {
            map = calls.getValue();
        }

        // finish activity if there are no calls
        if (map.isEmpty()) {
            displayState.setValue(new CallDisplayState(null, null));
            attemptFinishActivity();
            NotificationHelper.tearDown(callService);
            return;
        }

        // set primary and secondary calls
        OngoingCall primary = getPrimaryCallToDisplay();
        if (primary == null) {
            // Conference creation/removal is reported through several callbacks.
            // Keep a surviving call visible while its parent relationship settles
            // instead of briefly publishing an empty screen and closing the UI.
            primary = getPreviousVisibleCall(map);
        }
        if (primary == null) {
            primary = getFirstTrackedCall(map);
        }

        if (primary != null) {
            OngoingCall secondary = getCallToDisplay(primary);
            displayState.setValue(new CallDisplayState(primary, secondary));
            handleCallNotification(primary, primary.getState());
            if (primary.getState() == Call.STATE_DIALING) attemptStartActivity();
            updateProximitySensor(primary);
        } else {
            displayState.setValue(new CallDisplayState(null, null));
        }
    }

    private boolean reconcileCallsFromTelecom() {
        if (callService == null) return false;

        Map<Call, OngoingCall> reconciledCalls = new HashMap<>(calls.getValue());
        boolean changed = false;
        for (Call call : callService.getCalls()) {
            if (call.getState() != Call.STATE_DISCONNECTED && !reconciledCalls.containsKey(call)) {
                reconciledCalls.put(
                        call,
                        new OngoingCall(context, call, this, nextCallSequence++)
                );
                resolveContact(reconciledCalls.get(call));
                changed = true;
            }
        }

        if (!changed) return false;

        // Telecom can expose a new conference parent before onCallAdded reaches
        // us, or retain it after an independent companion call is removed.
        calls.setValue(reconciledCalls);
        return true;
    }

    @MainThread
    void onCallDetailsChanged(OngoingCall ongoingCall) {
        ongoingCall.refreshIdentity();
        resolveContact(ongoingCall);
        updateCalls();
    }

    private void resolveContact(OngoingCall ongoingCall) {
        String number = ongoingCall.getCallerNumber();
        if (number.isEmpty() || ongoingCall.isConference()) return;

        contactResolver.resolve(number, contact -> {
            Map<Call, OngoingCall> currentCalls = calls.getValue();
            if (currentCalls == null || currentCalls.get(ongoingCall.getCall()) != ongoingCall) return;
            if (!number.equals(ongoingCall.getCallerNumber())) return;

            ongoingCall.applyContact(contact);
            updateCalls();
        });
    }

    @Nullable
    private OngoingCall getPreviousVisibleCall(Map<Call, OngoingCall> map) {
        CallDisplayState previous = displayState.getValue();
        if (previous == null || previous.getPrimary() == null) return null;

        OngoingCall call = previous.getPrimary();
        return map.containsValue(call) && call.getState() != Call.STATE_DISCONNECTED ? call : null;
    }

    @Nullable
    private OngoingCall getFirstTrackedCall(Map<Call, OngoingCall> map) {
        OngoingCall first = null;
        for (OngoingCall current : map.values()) {
            if (current.getState() == Call.STATE_DISCONNECTED) continue;
            if (first == null || current.getSequence() < first.getSequence()) first = current;
        }
        return first;
    }

    private void handleCallNotification(OngoingCall call, int state) {
        if (callService == null) return;
        String caller = call.getCallerName();
        if (caller == null || caller.isEmpty()) caller = call.getCallerNumber();
        if (caller == null || caller.isEmpty()) caller = context.getString(R.string.anonymous);

        switch (state) {
            case Call.STATE_RINGING:
                if (!isActivityShowing())
                    NotificationHelper.notifyIncomingCall(context, callService, caller);
                break;
            case Call.STATE_DIALING:
                NotificationHelper.notifyOutgoingCall(context, callService, caller);
                break;
            case Call.STATE_ACTIVE:
                NotificationHelper.notifyOngoingCall(context, callService, caller);
                break;
            case Call.STATE_HOLDING:
                NotificationHelper.notifyOnHoldCall(context, callService, caller);
                break;
        }
    }

    private void updateProximitySensor(OngoingCall pCall) {
        OngoingCall call = pCall;
        if (call == null) {
            CallDisplayState currentDisplayState = displayState.getValue();
            call = currentDisplayState == null ? null : currentDisplayState.getPrimary();
        }

        if (proximitySensor == null
                || call == null
                || audioState.getValue() == null)
            return;

        int state = call.getState();
        int audioRoute = audioState.getValue().getRoute();
        proximitySensor.updateProximitySensorMode(state, audioRoute);
    }

    private OngoingCall getPrimaryCallToDisplay() {
        if (getFirstRingingCall() != null) return getFirstRingingCall();
        else if (getFirstDialingCall() != null) return getFirstDialingCall();
        else if (getFirstConnectingCall() != null) return getFirstConnectingCall();
        else if (getCallToDisplay(null) != null) return getCallToDisplay(null);
        return null;
    }

    private OngoingCall getCallToDisplay(OngoingCall ignore) {
        if (getFirstActiveCall() != null && getFirstActiveCall() != ignore)
            return getFirstActiveCall();
        else if (getFirstHoldingCall() != null && getFirstHoldingCall() != ignore)
            return getFirstHoldingCall();
        else if (getSecondHoldingCall() != null && getSecondHoldingCall() != ignore)
            return getSecondHoldingCall();
        else if (getFirstDisconnectingCall() != null && getFirstDisconnectingCall() != ignore)
            return getFirstDisconnectingCall();
        return null;
    }

    private OngoingCall getFirstConnectingCall() {
        return getFirstCallWithState(Call.STATE_CONNECTING);
    }

    private OngoingCall getFirstDialingCall() {
        return getFirstCallWithState(Call.STATE_DIALING);
    }

    private OngoingCall getFirstRingingCall() {
        return getFirstCallWithState(Call.STATE_RINGING);
    }

    private OngoingCall getFirstActiveCall() {
        return getFirstCallWithState(Call.STATE_ACTIVE);
    }

    private OngoingCall getFirstHoldingCall() {
        return getFirstCallWithState(Call.STATE_HOLDING);
    }

    private OngoingCall getSecondHoldingCall() {
        return getSecondCallWithState(Call.STATE_HOLDING);
    }

    private OngoingCall getFirstDisconnectingCall() {
        return getFirstCallWithState(Call.STATE_DISCONNECTING);
    }

    @Nullable
    private OngoingCall getFirstCallWithState(int state) {
        if (calls.getValue() == null) return null;

        OngoingCall first = null;
        for (OngoingCall current : calls.getValue().values()) {
            if (current.getState() == state && !current.isConferenced()) {
                if (first == null || current.getSequence() < first.getSequence()) {
                    first = current;
                }
            }
        }

        return first;
    }

    @Nullable
    private OngoingCall getSecondCallWithState(int state) {
        if (calls.getValue() == null) return null;
        OngoingCall first = null;
        OngoingCall second = null;

        for (OngoingCall current : calls.getValue().values()) {
            if (current.getState() == state && !current.isConferenced()) {
                if (first == null || current.getSequence() < first.getSequence()) {
                    second = first;
                    first = current;
                } else if (second == null || current.getSequence() < second.getSequence()) {
                    second = current;
                }
            }
        }

        return second;
    }

    public void attemptFinishActivity() {
        if (isActivityStarted()) {
            inCallActivity.finish();
        }
    }

    public void attemptStartActivity() {
        if (!isActivityShowing() && context != null) {
            InCallActivity.Companion.start(context);
        }
    }

    @MainThread
    public void updateCallAudioState(CallAudioState newAudioState) {
        audioState.setValue(new CallAudioUiState(newAudioState.getRoute(), newAudioState.isMuted()));
        updateProximitySensor(null);
    }

    @MainThread
    public void updateCallEndpoint(int route) {
        CallAudioUiState current = audioState.getValue();
        audioState.setValue(new CallAudioUiState(route, current != null && current.isMuted()));
        updateProximitySensor(null);
    }

    @MainThread
    public void updateMuteState(boolean isMuted) {
        CallAudioUiState current = audioState.getValue();
        int route = current == null ? CallAudioState.ROUTE_EARPIECE : current.getRoute();
        audioState.setValue(new CallAudioUiState(route, isMuted));
    }

    @MainThread
    public void updateCanAddCall(boolean newCanAddCall) {
        canAddCall.setValue(newCanAddCall);
    }


    public void setup(InCallServiceImpl callService, Context context, ProximitySensor proximitySensor) {
        this.callService = callService;
        this.context = context;
        this.proximitySensor = proximitySensor;
        updateCalls();
    }

    public void tearDown() {
        callService = null;
        context = null;
        if (proximitySensor != null) proximitySensor.tearDown();
        proximitySensor = null;
    }

    public LiveData<CallAudioUiState> getAudioState() {
        return audioState;
    }

    public LiveData<Boolean> getCanAddCall() {
        return canAddCall;
    }

    public LiveData<CallDisplayState> getDisplayState() {
        return displayState;
    }

    public LiveData<Map<Call, OngoingCall>> getCalls() {
        return calls;
    }
}

File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/CallsHandler.kt
package dev.alenajam.opendialer.feature.inCall.service

import android.content.Context
import android.telecom.Call
import android.telecom.CallAudioState
import android.telecom.VideoProfile
import androidx.annotation.MainThread
import dev.alenajam.opendialer.feature.inCall.ui.InCallActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CallsHandler @Inject constructor(
    private val contactResolver: CallContactResolver,
    private val telecomAdapter: TelecomAdapter
) : CallManager {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val _calls = MutableStateFlow<Map<Call, OngoingCall>>(emptyMap())
    override val calls: StateFlow<Map<Call, OngoingCall>> = _calls.asStateFlow()

    @OptIn(ExperimentalCoroutinesApi::class)
    override val displayState: StateFlow<CallDisplayState> = _calls
        .flatMapLatest { map ->
            if (map.isEmpty()) {
                flowOf(CallDisplayState(null, null))
            } else {
                combine(map.values.map { it.stateFlow }) { _ ->
                    deriveDisplayState(map)
                }
            }
        }
        .stateIn(scope, SharingStarted.Eagerly, CallDisplayState(null, null))

    private val _audioState = MutableStateFlow<CallAudioUiState?>(null)
    override val audioState: StateFlow<CallAudioUiState?> = _audioState.asStateFlow()

    private val _canAddCall = MutableStateFlow(false)
    override val canAddCall: StateFlow<Boolean> = _canAddCall.asStateFlow()

    private val _events = MutableSharedFlow<CallEvent>(extraBufferCapacity = 1)
    override val events: SharedFlow<CallEvent> = _events.asSharedFlow()

    private var callService: InCallServiceImpl? = null
    private var context: Context? = null
    private var nextCallSequence: Long = 0

    init {
        // Observe displayState to handle side-effects like Activity start/finish
        displayState.onEach { state ->
            if (state.primary == null && _calls.value.isEmpty()) {
                _events.tryEmit(CallEvent.FinishActivity)
            } else if (state.primary?.state == Call.STATE_DIALING) {
                attemptStartActivity()
            }
        }.launchIn(scope)
    }

    private fun deriveDisplayState(map: Map<Call, OngoingCall>): CallDisplayState {
        val selection = selectDisplayCalls(map)
        val primary = selection.primary ?: getPreviousVisibleCall(map) ?: getFirstTrackedCall(map)
        return CallDisplayState(
            primary = primary,
            primaryState = primary?.stateFlow?.value,
            secondary = selection.secondary,
            secondaryState = selection.secondary?.stateFlow?.value
        )
    }

    // CallManager Implementation (Delegated Actions)
    override fun answer(call: OngoingCall) {
        call.answer()
    }

    override fun hangup(call: OngoingCall, message: String?) {
        call.hangup(message)
    }

    override fun hold(call: OngoingCall, hold: Boolean?) {
        call.hold(hold)
    }

    override fun startDtmf(call: OngoingCall, digit: Char) {
        call.startDtmf(digit)
    }

    override fun stopDtmf(call: OngoingCall) {
        call.stopDtmf()
    }

    override fun merge(call: OngoingCall) {
        call.merge()
    }

    override fun split(call: OngoingCall) {
        call.split()
    }

    override fun swap() {
        val secondary = displayState.value.secondary ?: return
        hold(secondary, false)
    }

    // Audio Commands (Delegated to TelecomAdapter)
    override fun toggleSpeaker() = telecomAdapter.toggleSpeaker()
    override fun toggleBluetooth() = telecomAdapter.toggleBluetooth()
    override fun selectAudioRoute(route: CallAudioRouteUiState) = telecomAdapter.selectAudioRoute(route)
    override fun toggleMute() = telecomAdapter.toggleMute()

    // Lifecycle and Event Methods
    @MainThread
    fun addCall(call: Call, context: Context) {
        if (call.safeState == Call.STATE_DISCONNECTED) {
            OngoingCallHelper.handleDisconnectCause(context, call)
            return
        }

        var newOngoingCall: OngoingCall? = null
        _calls.update { currentMap ->
            if (currentMap.containsKey(call)) {
                currentMap[call]?.refreshIdentity()
                currentMap
            } else {
                val ongoingCall = OngoingCall(
                    context,
                    call,
                    onRemoved = { removeCall(it) },
                    sequence = nextCallSequence++
                )
                newOngoingCall = ongoingCall
                currentMap + (call to ongoingCall)
            }
        }

        newOngoingCall?.let { resolveContact(it) }
        reconcileCallsFromTelecom()
    }

    @MainThread
    fun removeCall(call: Call) {
        var removedCall: OngoingCall? = null
        _calls.update { currentMap ->
            removedCall = currentMap[call]
            currentMap - call
        }

        removedCall?.let { ongoingCall ->
            if (ongoingCall.shouldNotifyMissedCall) {
                _events.tryEmit(
                    CallEvent.MissedCall(
                        callerName = ongoingCall.callerName,
                        callerNumber = ongoingCall.callerNumber,
                        notificationId = ((System.currentTimeMillis() + ongoingCall.sequence) and Int.MAX_VALUE.toLong()).toInt()
                    )
                )
            }
            ongoingCall.tearDown()
        }
        reconcileCallsFromTelecom()
    }

    private fun reconcileCallsFromTelecom(): Boolean {
        val service = callService ?: return false
        var changed = false

        val callsToResolve = mutableListOf<OngoingCall>()

        _calls.update { currentMap ->
            val reconciledMap = currentMap.toMutableMap()
            var mapChanged = false

            fun addIfNeeded(call: Call) {
                if (call.safeState != Call.STATE_DISCONNECTED && !reconciledMap.containsKey(call)) {
                    val ongoingCall = OngoingCall(
                        context!!,
                        call,
                        onRemoved = { removeCall(it) },
                        sequence = nextCallSequence++
                    )
                    reconciledMap[call] = ongoingCall
                    callsToResolve.add(ongoingCall)
                    mapChanged = true
                }
            }

            for (call in service.calls) {
                addIfNeeded(call)
                for (child in call.children) {
                    addIfNeeded(child)
                }
            }

            if (mapChanged) {
                changed = true
                reconciledMap
            } else {
                currentMap
            }
        }

        callsToResolve.forEach { resolveContact(it) }
        return changed
    }

    private fun resolveContact(ongoingCall: OngoingCall) {
        val number = ongoingCall.callerNumber
        if (number.isEmpty() || ongoingCall.isConference) return
        scope.launch {
            val contact = contactResolver.resolve(number)
            if (_calls.value[ongoingCall.call] !== ongoingCall) return@launch
            if (number != ongoingCall.callerNumber) return@launch
            ongoingCall.applyContact(contact)
        }
    }

    private fun getPreviousVisibleCall(map: Map<Call, OngoingCall>): OngoingCall? {
        val previous = displayState.value.primary ?: return null
        return if (map.containsValue(previous) && previous.state != Call.STATE_DISCONNECTED) previous else null
    }

    private fun getFirstTrackedCall(map: Map<Call, OngoingCall>): OngoingCall? {
        return map.values.filter { it.state != Call.STATE_DISCONNECTED }.minByOrNull { it.sequence }
    }

    private fun selectDisplayCalls(map: Map<Call, OngoingCall>): CallDisplaySelector.Selection<OngoingCall> {
        val candidates = map.values.map {
            CallDisplaySelector.Candidate(it, it.state, it.isConferenced, it.sequence)
        }
        return CallDisplaySelector.select(candidates)
    }

    fun setup(callService: InCallServiceImpl, context: Context) {
        this.callService = callService
        this.context = context
        reconcileCallsFromTelecom()
    }

    fun tearDown() {
        _calls.value.values.forEach { it.tearDown() }
        _calls.value = emptyMap()
        _audioState.value = null
        _canAddCall.value = false
        callService = null
        context = null
    }

    @MainThread
    fun updateCallAudioState(newAudioState: CallAudioState) {
        val routes = listOf(
            CallAudioState.ROUTE_EARPIECE,
            CallAudioState.ROUTE_SPEAKER,
            CallAudioState.ROUTE_BLUETOOTH,
            CallAudioState.ROUTE_WIRED_HEADSET
        ).filter { newAudioState.supportedRouteMask and it != 0 }
            .map { route ->
                CallAudioRouteUiState(
                    type = route,
                    label = TelecomAdapter.defaultRouteLabel(route),
                    isSelected = route == newAudioState.route
                )
            }
        _audioState.value = CallAudioUiState(newAudioState.route, newAudioState.isMuted, routes)
    }

    @MainThread
    fun updateAvailableAudioRoutes(route: Int, routes: List<CallAudioRouteUiState>) {
        _audioState.update { current ->
            CallAudioUiState(
                route = route,
                isMuted = current?.isMuted == true,
                availableRoutes = routes
            )
        }
    }

    @MainThread
    fun updateCallEndpoint(route: Int) {
        val isMuted = _audioState.value?.isMuted == true
        _audioState.value = CallAudioUiState(route, isMuted)
    }

    @MainThread
    fun updateMuteState(isMuted: Boolean) {
        val route = _audioState.value?.route ?: CallAudioState.ROUTE_EARPIECE
        _audioState.value = CallAudioUiState(route, isMuted)
    }

    @MainThread
    fun updateCanAddCall(newCanAddCall: Boolean) {
        _canAddCall.value = newCanAddCall
    }

    fun attemptStartActivity() {
        context?.let { InCallActivity.start(it) }
    }

}
